dartagnan/src/main/java/com/dat3m/dartagnan/program/event/AbstractEvent.java

```diff
@@ -217,20 +218,35 @@ public void insertAfter(List<Event> toBeInserted) {
     }
 
     @Override
-    public void replaceBy(Event replacement) {
+    public void insertBefore(Event toBeInserted) {
+        Preconditions.checkNotNull(toBeInserted);
+        insertBetween((AbstractEvent) toBeInserted, function, this.predecessor, this);
+    }
+
+    @Override
+    public void insertBefore(Iterable<? extends Event> toBeInserted) {
+        for (Event next : toBeInserted) {
+            this.insertBefore(next);
+        }
+    }
+
+    @Override
+    public Event replaceBy(Event replacement) {
         if (replacement == this) {
-            return;
+            return replacement;
         }
         Preconditions.checkState(currentUsers.isEmpty(), "Cannot replace event that is still in use.");
         this.insertAfter(replacement);
         this.forceDelete();
+        return replacement;
     }
 
     @Override
-    public void replaceBy(List<Event> replacement) {
+    public <TCol extends Iterable<? extends Event>> TCol replaceBy(TCol replacement) {
```

Comment:
What is this `TCol`?

Reply:
Short for `TCollection` (thought technically it's an `Iterable`).
The idea was that you can call something like
```
List<Event> replacement = event.replaceBy(eventSequence(...));
```
It just returns the collection of events you have put in with the same type (so if you used `Set<Event>` it would return a set).
That being said, I didn't end up using it anywhere I believe. So I might remove this feature again.

Reply:
I removed the return values again.